### PR TITLE
Remove long double instantiations.

### LINKED ISSUE #2551: Problems with Intel compiler and complex long double
When trying to compile the library with the Intel compiler, version 16, I get the (many) of the following erros:
```
/home/kronbichler/sw/deal.II/include/deal.II/lac/constraint_matrix.templates.h(1604): error: no operator "*" matches these operands
            operand types are: const std::complex<long double> * double
            col_val += (local_matrix(loc_row, global_cols.local_row(j,p)) *
                                                                          ^
[...]
            instantiation of "void dealii::ConstraintMatrix::distribute_local_to_global(const dealii::FullMatrix<MatrixType::value_type> &, const dealii::Vector<VectorType::value_type> &,
            const std::vector<unsigned int, std::allocator<unsigned int>> &, MatrixType &, VectorType &, bool, dealii::internal::bool2type<false>) const
            [with MatrixType=dealii::SparseMatrix<std::complex<long double>>, VectorType=dealii::Vector<std::complex<long double>>]" at line 1305 of
                      "/home/kronbichler/sw/deal.II/source/lac/constraint_matrix.cc"
```
The problem is that the Intel compiler does not seem to understand multiplication between `std::complex<long double>` and `double`. Has anyone else seen this? My solution would be to get rid of `std::complex<long double>` instantiations, both in special files and as a complex scalar. I am not really sure in which context they might be of use anyway. The trend seems to go away from long double (if there ever was) because the 80 bit long double is getting more of a niche as it is the only format left on the old x87 format.

In case my system is weird, here are my compiler details:
```
icpc version 16.0.1 (gcc version 4.8.5 compatibility)
```
(I tried to attach gcc 5 to the Intel compiler but I get many C++11-related errors, it looks like Intel is not ready for that yet.)

### PR DESCRIPTION
**Pull request by kronbichler**
This fixes #2551.

It was relatively straight-forward to get rid of long double in the vector and matrix types. The reduction is also beneficial for library sizes. The debug library went for 1.22 GB to 1.03 GB and the release library from 149 MB to 124 MB, both a reduction of almost 20%.

I left some instatiations of `FullMatrix<long double>` and `Vector<long double>` (base variants) in place because there is some use in mapping_q_generic but I'm also willing to change that to a double matrix (it is used for the initial computation of the Laplace vector, i.e., support points on quad/hex). Not sure if the additional precision is really necessary.

As far as I can see, this patch passes the test suite (it is a bit difficult to find failing tests on the command line when there are around 20 other failing tests)...

(Do not merge yet; I need to write a changelog/incompatibility entry and want to leave some time for others to comment.)

### CODE REVIEW

**Comment by bangerth** [comment]
For the failing tests, yes, see #2530 :-(

File: include/deal.II/lac/vector.h [hunk 1]
```diff
@@ -104,11 +104,9 @@ struct VectorOperation
...
  *
  * @note Instantiations for this template are provided for <tt>@<float@>,
- * @<double@>, @<long double@>, @<std::complex@<float@>@>,
- * @<std::complex@<double@>@>, @<std::complex@<long double@>@></tt>; others
- * can be generated in application programs (see the section on
- * @ref Instantiations
- * in the manual).
+ * @<double@>, @<std::complex@<float@>@>, @<std::complex@<double@>@></tt>;
+ * others can be generated in application programs (see the section on @ref
+ * Instantiations in the manual).
```
Comment (bangerth) [inline include/deal.II/lac/vector.h]: Please move the `@ref` on the same line as the tag.

File: include/deal.II/lac/vector_view.h [hunk 2]
```diff
@@ -120,11 +120,9 @@ DEAL_II_NAMESPACE_OPEN
...
  *
  * @note Instantiations for this template are provided for <tt>@<float@>,
- * @<double@>, @<long double@>, @<std::complex@<float@>@>,
- * @<std::complex@<double@>@>, @<std::complex@<long double@>@></tt>; others
- * can be generated in application programs (see the section on
- * @ref Instantiations
- * in the manual).
+ * @<double@>, @<std::complex@<float@>@>, @<std::complex@<double@>@></tt>;
+ * others can be generated in application programs (see the section on @ref
+ * Instantiations in the manual).
```
Comment (bangerth) [inline include/deal.II/lac/vector_view.h]: same here.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 3]
```diff
@@ -7085,10 +7085,8 @@ namespace VectorTools
         // if v is not a boring Vector or BlockVector:
         Assert(   dynamic_cast<Vector<double> *>(& v)
                   || dynamic_cast<Vector<float> *>(& v)
-                  || dynamic_cast<Vector<long double> *>(& v)
                   || dynamic_cast<BlockVector<double> *>(& v)
-                  || dynamic_cast<BlockVector<float> *>(& v)
-                  || dynamic_cast<BlockVector<long double> *>(& v),
+                  || dynamic_cast<BlockVector<float> *>(& v),
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: I suspect that the change doesn't hurt -- but the code probably wouldn't be wrong with the old code either. Opinions?

**Comment by bangerth** [comment]
Looks good to me up to the couple of comments.

Please merge at your discretion when ready.

File: include/deal.II/numerics/vector_tools.templates.h (hunk 3 above)
Comment (kronbichler) [inline include/deal.II/numerics/vector_tools.templates.h]: You are right, we can still leave it in here. Thanks for spotting this.

**Comment by tjhei** [comment]
Looks good on my end too.

**Comment by kronbichler** [comment]
I re-checked the tests. On my machine there are 27 failures. But those are all known to me from before. I've removed the WIP label but let someone else do the merge.

**Pull request merged**